### ISSUE TITLE
PayPal and Magento2 2.3.1

### ISSUE THREAD

**Issue by chequille** [magento/magento2#22528]
Hi,
it is really great, that everytime there is an update you have to do all work from the beginning again !!!!

Now, as I had PayPal working fine (in 2.3.0) with a button "zahlungspflicht bestellen" (for those not knowing this it is mandatory in Germany), there is a new paypal module doing it completly different.

Does somebody cares about German Stores? Sometimes I think no.

Therefore, any ideas, hints how I can get back my own button, not the one in the inserted iframe?

As well, how can I get rid of the second button displayed which offers "Ratenzahlung". Do not need it and do not whant to have it.

I guess there are many people out there having the same issue. Hopefully somebody can help.

Thanks for reading this.
Best regards,
Jürgen

**Comment by chequille** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @chequille. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @chequille, here is your [Magento instance](https://i-22528-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-22528-2-3-develop.instances.magento-community.engineering/admin](https://i-22528-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by chequille** [comment]
I confirm, that vanilla Magento instance does have the same problem. 
PayPal Credit Button is shown but it should not as of the describtions of the refund options.
It says "PayPal credit is only availabe in those countries where it is allowed and the currency is USD"
But it is displayed in the front end.
In the backend settings of PayPal "Disable Funding Options", it is not shown as it should.
Where is the problem?
Please advise

**Comment by stkristobal** [comment]
I have a similar issue with the PP module; I get the regular button, but selecting the PP option and clicking 'Place Order' simply does nothing. This arose after 2.3.1.
There's no log stating an error anywhere; but I do get this in the Chrome output:

"Error: Expected props.locale to be valid
    at Object.validateButtonLocale [as validate] (https://www.paypalobjects.com/api/checkout.js:13198:101)
    at validateProp (https://www.paypalobjects.com/api/checkout.js:6884:73)
    at https://www.paypalobjects.com/api/checkout.js:7272:34
    at ParentComponent.setProps (https://www.paypalobjects.com/api/checkout.js:7278:18)
    at new ParentComponent (https://www.paypalobjects.com/api/checkout.js:7044:27)
    at https://www.paypalobjects.com/api/checkout.js:6568:28
    at Function.ZalgoPromise.try (https://www.paypalobjects.com/api/checkout.js:5660:37)
    at Component.render (https://www.paypalobjects.com/api/checkout.js:6567:47)
    at callOriginal (https://www.paypalobjects.com/api/checkout.js:17111:41)
    at https://www.paypalobjects.com/api/checkout.js:14531:20"

Maybe not 100% relevant, but I have no idea what's causing this - hoping it's slightly related to this,.

**Comment by neeta-wagento** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by mikedubinsky** [comment]
@chequille I recently posted on [stackexchange](https://magento.stackexchange.com/questions/273398/magento-2-3-1-changed-paypal-express-button) about a similar issue that we've experienced with Magento 2.3.1 changing our PayPal button. Is your button implemented in a similar way?

The Magento 2.3.1 upgrade changed our PayPal Express button which we have configured to be displayed on the Product page through:

`app/design/frontend/theme/default/Magento_Paypal/templates/express/shortcut.phtml`

where we have modified the image

`...src="<?=` $this->getViewFileUrl('images/paypalicon.png'); ?>"...`

I think this is a result of the 2.3.1 PayPal changes and Smart Buttons that we'd like to disable, but we already have Customize set to No in PayPal > Frontend > Product Page

You can see the change below:
![image](https://user-images.githubusercontent.com/44975166/57171948-bf8be480-6de7-11e9-9668-c028689db089.png)

Thanks for your help and collaboration all.

**Comment by chequille** [comment]
@mikedubinsky : First of all thanks for this input.
But it is not the same problem as i have on my system.
On my site, the paypal stuff seems to be implement now with 2.3.1 completly different as before.
I can see an iframe which is loaded and the buttons for PayPal are defined there. The mentioned template for the shortcuts are not used at all.
The component which seems to be loaded is named "zoid-component" or something like that.
This seems to be a completly other implementation now. The only influene on the buttons is to set some special things in the backend configuration pages, but the main buttons remain as this module implements it.
I do not see any way to get the old stuff activated.

What I tried is to replace the complete folder "magento-paypal" in vendor/magento with the version from 2.3.0 and it works as before. 
I have my button displayed again and as well the addition button for "PayPal Credit" is not shown.

That is the only way currently to get PayPal working as I want on my site.

Regards,
Jürgen

**Comment by chequille** [comment]
Some additional information maybe for somebody taking care about this problem:
My shop is set to country Germany and currency isi of course EURO.
No other currencies allowed and no other countries alloewd.

**Comment by chequille** [comment]
@mikedubinsky
Mike, have a look into your website. I guess you have the same problem now. There is an iframe loaded which is showing the buttons.
Regards
Jürgen

**Comment by mikedubinsky-odb** [comment]
@chequille Yes, instead of our custom 'PayPal' button, we now see a 'zoid-paypal-button-...' iframe. @neeta-wagento, was this issue something that you also experienced? Thanks all!

**Comment by chequille** [comment]
Yes, exactly this is my problem. I need a button where i can clearly define my own text "zahlungpflichtig bestellen".
I had that with Magento 2.3.0
But now the buttons and the texts on it are made from paypal and you do not have a real influence on what is written on it. Only standard labels like "Checkout" or "PayPal2 etc. are possible.
I really think about installing the modul from 2.3.0, because there it works as I want it.
THe only question is how long it will work. Mabe with the next update of Magento it is not working anymore.
I am really a little bit angry about this, because PayPal and a little bit Magento as well is dictating you what can be done and what cannot be done.
Ok, will see if somebody here officillay will make a comment on our issue here. Till now nobody answered.

See you
BR
Jürgen

**Comment by therool** [comment]
I have encountered the same issue.
The issue I see is this:
- [Magento\Paypal\Block\Adminhtml\System\Config\MultiSelect\DisabledFundingOptions@filterValuesForPaypalCredit](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Paypal/Block/Adminhtml/System/Config/MultiSelect/DisabledFundingOptions.php#L72) which removes the "CREDIT" from the option list so we can't disable it in admin area - don't know what is the point of removing it if by default it is included in code regardless of the merchant country
- By default the "CREDIT" is always allowed because the method [Magento\Paypal\Model\SmartButtonConfig@getAllowedFunding](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Paypal/Model/SmartButtonConfig.php#L94) will never filter out "CREDIT" as not available because it is passed in the constructor [Magento/Paypal/etc/frontend/di.xm](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Paypal/etc/frontend/di.xml#L160) regardless of the merchant country and can't be removed because the option to do so it removed by the issue mentioned above

To fix this create an custom module which depends on "Magento_Paypal" and create an file `etc/frontend/di.xml` file with the following contents (Change the namespace and location accordingly your needs):
```xml
<?xml version="1.0"?>
<!--
/**
 * @category    Monobunt
 * @package     Monobunt_Ammerer
 * @author      Ricards Zalitis <info@monobunt.at>
 * @copyright   2019 monobunt.at
 * @license     Unlimited Commercial License
 */
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <!--Attach "CREDIT" as disallowed funding if the paypal merchant country is not equal to US-->

    <type name="Magento\Paypal\Model\SmartButtonConfig">
        <plugin name="Monobunt_Ammerer::SmartButtonConfigPlugin"
                type="Monobunt\Ammerer\Plugin\Magento\Paypal\Model\SmartButtonConfigPlugin"
                disabled="false"/>
    </type>

    <!--Take in account the "Display on Product Details Page" setting-->
    <type name="Magento\Paypal\Block\Express\InContext\SmartButton">
        <plugin name="Monobunt_Ammerer::SmartButtonPlugin"
                type="Monobunt\Ammerer\Plugin\Magento\Paypal\Block\Express\InContext\SmartButtonPlugin"
                disabled="false"/>
    </type>
</config>
```
Create an file `SmartButtonConfigPlugin.php` which will remove the "CREDIT" from allowed fundings and add it to disallowed funding if the merchant country is not US:
```php
<?php
/**
 * @category    Monobunt
 * @package     Monobunt_Ammerer
 * @author      Ricards Zalitis <info@monobunt.at>
 * @copyright   2019 monobunt.at
 * @license     Unlimited Commercial License
 */

namespace Monobunt\Ammerer\Plugin\Magento\Paypal\Model;

use Magento\Paypal\Model\{
    Config,
    ConfigFactory
};
use Magento\Paypal\Model\SmartButtonConfig;

class SmartButtonConfigPlugin
{
    /**
     * @var Config
     */
    private $config;

    /**
     * SmartButtonConfigPlugin constructor
     *
     * @param ConfigFactory $configFactory
     */
    public function __construct(ConfigFactory $configFactory)
    {
        $this->config = $configFactory->create();
        $this->config->setMethod(Config::METHOD_EXPRESS);
    }

    /**
     * Attach "CREDIT" as disallowed funding if the paypal merchant country is not equal to US
     *
     * @param SmartButtonConfig $subject
     * @param $result
     * @return array
     */
    public function afterGetConfig(SmartButtonConfig $subject, $result): array
    {
        if ($this->config->getMerchantCountry() != 'US') {
            $creditIndex = array_search('CREDIT', $result['allowedFunding']);

            if ($creditIndex !== false) {
                $result['disallowedFunding'][] = 'CREDIT';
                array_splice($result['allowedFunding'], $creditIndex, 1);
            }
        }

        return $result;
    }
}
```
Optionaly create an file `SmartButtonPlugin.php` so the "Display on Product Details Page" setting is correctly taken in account:
```php
<?php
/**
 * @category    Monobunt
 * @package     Monobunt_Ammerer
 * @author      Ricards Zalitis <info@monobunt.at>
 * @copyright   2019 monobunt.at
 * @license     Unlimited Commercial License
 */

namespace Monobunt\Ammerer\Plugin\Magento\Paypal\Block\Express\InContext;

use Magento\Paypal\Model\{
    Config,
    ConfigFactory
};
use Magento\Paypal\Block\Express\InContext\SmartButton;

class SmartButtonPlugin
{
    /**
     * @var Config
     */
    private $config;

    /**
     * SmartButtonConfigPlugin constructor
     *
     * @param ConfigFactory $configFactory
     */
    public function __construct(ConfigFactory $configFactory)
    {
        $this->config = $configFactory->create();
        $this->config->setMethod(Config::METHOD_EXPRESS);
    }

    /**
     * Take in account the "Display on Product Details Page" setting
     *
     * @param SmartButton $subject
     * @param callable $proceed
     * @param string $key
     * @param null $index
     * @return bool
     */
    public function aroundGetData(SmartButton $subject, callable $proceed, $key = '', $index = null)
    {
        $result = $proceed($key, $index);

        if ($key === 'is_in_catalog_product') {
            return $result && $this->config->getValue('visible_on_product');
        }

        return $result;
    }
}
```

**Comment by luoine** [comment]
I use magento2.3.1 and install Chinese language pack (taiwan), I set the language locale is English (United States) the paypal work fine, While I change the locale to Chinese (Taiwan) the paypal button is not shown.

**Comment by ghost** [comment]
Looks like a bug.
`/vendor/magento/module-paypal/Model/AbstractConfig.php`
Instead of
`'payment/paypal_express/disable_funding_options',`
should be
`'paypal/style/disable_funding_options'`

**Comment by chequille** [comment]
@mikedubinsky
Hi Mike,
I do not know whether you still have the problem with your PayPal buttons, but if yes, I guess I know why.

Configuring PayPal in the backend "not using InContext Experience" does not work, because if you set it to No, the value is not stored under skope "websites", it is stored under skope "default".
Therefore in the frontend you will get always the iframe included which of course does not show your buttons.

Have a look at issue #24144
I described it there what to set in core_config_data table manually.

Unfortunatley, nobody cares and the issue was closed. But it is a bug as far as I see it.

Regards,
Chequille

**Comment by oleksii-lisovyi** [comment]
Hi, @chequille!

I've created a PR #24121 , that was already tested and accepted. So sooner on later it will be released with upcoming Magento 2.3.x version.

For now you can apply patch from the commit https://github.com/magento/magento2/pull/24121/commits/7507dcef7dfa144a59df3e134ad3819963fc05d8

Here is an article how to apply patch for Magento:
https://devdocs.magento.com/guides/v2.3/comp-mgr/patching.html

**Comment by magento-engcom-team** [comment]
Hi @chequille. Thank you for your report. 
The issue has been fixed in magento/magento2#24121 by @oleksii-lisovyi in 2.3-develop branch
Related commit(s):
 - [7507dcef7dfa144a59df3e134ad3819963fc05d8](https://github.com/magento/magento2/commit/7507dcef7dfa144a59df3e134ad3819963fc05d8)
 - [c80bcb929325d2a3a4319d9b17d75a73c5731afe](https://github.com/magento/magento2/commit/c80bcb929325d2a3a4319d9b17d75a73c5731afe)
 - [ce7b33c6e1c1d416ecbd852b09a4d6a384394c7c](https://github.com/magento/magento2/commit/ce7b33c6e1c1d416ecbd852b09a4d6a384394c7c)

The fix will be available with the upcoming 2.3.4 release.
